### PR TITLE
Suggested example for PER.4

### PR DESCRIPTION
**Pull request by Eliyahu-Ravuna**
Don't assume that complicated code is necessarily faster than simple code.

The suggested example has been tested with GCC 4.9.2 with the -O3 flag and with MSVC 19.00 with the /Ox flag.

* The simple version takes about 35 - 70 ms depending on the compiler.
* The complicated version with handcrafted optimization takes about 130 ms.

The simple code is faster because most modern compilers will translate it to SIMD instructions via automatic vectorization.

The full code of the example is below.
(given here for the comment of the pull request only, not for the guidelines. The example suggested for the guidelines is a subset of this code)

    #include <iostream>
    #include <functional>
    #include <chrono>
    #include <vector>
    #include <string>
    
    using namespace std;
    using namespace std::chrono;
    
    vector<char> v(100000);
    
    void measure_time(const string& title, function<void()> f)
    {
        constexpr int times_to_test = 10000;
        auto start = system_clock::now();
        for(int i=0; i<times_to_test; ++i)
            f();
        auto duration = system_clock::now() - start;
        cout << title << ": " << duration_cast<milliseconds>(duration).count() << " milliseconds \n";
    }
    
    void good_example()
    {
        for(auto& c : v)
            c = ~c;
    }
    
    void bad_example()
    {    
        for(size_t i=0; i<v.size(); i+=sizeof(uint64_t))
        {
            uint64_t& quad_word = *reinterpret_cast<uint64_t*>(&v[i]);
            quad_word = ~quad_word;
        }    
    }
    
    int main()
    {  
      //cout << "Tested with GCC version " << __VERSION__ << '\n';
      //cout << "Tested with MSVC version " << _MSC_VER << '\n';

      constexpr int times_to_repeat = 3;
      for(int i=0; i<times_to_repeat; ++i)
      {
          measure_time("good_example", good_example);
          measure_time("bad_example", bad_example);
      }
    }

This is the output with GCC 4.9.2:

    Tested with GCC version 4.9.2
    good_example: 76 milliseconds 
    bad_example: 136 milliseconds 
    good_example: 68 milliseconds 
    bad_example: 135 milliseconds 
    good_example: 69 milliseconds 
    bad_example: 132 milliseconds

And the difference is even larger with MSVC 19.00:

    Tested with MSVC version 1900
    good_example: 35 milliseconds
    bad_example: 128 milliseconds
    good_example: 34 milliseconds
    bad_example: 127 milliseconds
    good_example: 35 milliseconds
    bad_example: 127 milliseconds

This reinforces the idea that handcrafted optimizations trying to outsmart the compiler hurt more than they help. The version with clear expression of the intent is not only more understandable but also faster.

Eliyahu Ravuna

### CODE REVIEW

**Comment by Eliyahu-Ravuna** [comment]
The difference is even larger with MSVC 19.00.

    Tested with MSVC version 1900
    good_example: 35 milliseconds
    bad_example: 128 milliseconds
    good_example: 34 milliseconds
    bad_example: 127 milliseconds
    good_example: 35 milliseconds
    bad_example: 127 milliseconds

This reinforces the idea that handcrafted optimizations trying to outsmart the compiler hurt more than they help.

Eliyahu Ravuna

**Comment by Eliyahu-Ravuna** [comment]
Yes, I agree.
I incorporated your feedback in the pull request.

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -9116,6 +9116,27 @@ If your program spends most of its time waiting for the web or for a human, opti
 
 Simple code can be very fast. Optimizers sometimes do marvels with simple code
 
+##### Example, good
+
+    // clear expression of intent, fast execution
+    
+    vector v(100000);
```
Comment (tkoeppe) [inline CppCoreGuidelines.md]: This `vector` is missing template arguments?

**Comment by Eliyahu-Ravuna** [comment]
Fixed in pull request #402.

**Pull request merged**